elasticdl/python/master/k8s_instance_manager.py

```diff
@@ -236,49 +262,14 @@ def _event_cb(self, event):
                 == "OOMKilled"
             ):
                 self._failed_pods.append(pod_name)
-                failed_pod = True
-                logger.info("Pod %s is OOMKilled." % pod_name)
-            if pod_name in self._worker_pod_name_to_id:
-                worker_id = self._worker_pod_name_to_id.get(pod_name)
-                self._worker_pods_phase[worker_id] = (pod_name, phase)
-                if evt_type == "DELETED" or failed_pod:
-                    del self._worker_pods_phase[worker_id]
-                    del self._worker_pod_name_to_id[pod_name]
-                    self._task_d.recover_tasks(worker_id)
-
-                    # If a deleted pod was not "Succeeded", relaunch a worker.
-                    relaunch_worker = (
-                        self._relaunch_deleted_live_worker
-                        and phase != "Succeeded"
-                    )
-
-            elif pod_name in self._ps_pod_name_to_id:
-                ps_id = self._ps_pod_name_to_id.get(pod_name)
-                self._ps_pods_phase[ps_id] = (pod_name, phase)
-                if evt_type == "DELETED" or failed_pod:
-                    del self._ps_pods_phase[ps_id]
-                    del self._ps_pod_name_to_id[pod_name]
-                    relaunch_ps = self._relaunch_deleted_live_ps
-            else:
-                logger.error("Unknown pod name: %s" % pod_name)
-                return
-
-        if relaunch_worker and worker_id >= 0:
-            logger.info("Relaunching worker.")
-            new_worker_id = self._next_worker_id()
-            self._start_worker(new_worker_id)
-            self._update_addr(
-                worker_id,
-                new_worker_id,
-                self._worker_addrs,
-                addr_get_fn=self._k8s_client.get_worker_service_address,
-            )
-        elif relaunch_ps:
-            logger.info("Relaunching ps.")
-            # Note: the ID and service address for relaunched parameter
-            # server are intentionally left unchanged to support fault
-            # tolerance.
-            self._start_ps(ps_id)
+                pods_phase[pod_name][1] = phase
+                if relaunch_flag:
+                    remove_fn(pods_phase[pod_name][0])
+                    relaunch_fn(pod_name)
```

Comment:
If the pod status is `OOMKilled`, we should kill it first, and then relaunch a new one.

Reply:
If we kill OOMKilled pods, the logs are lost. We have two choices: keep failed pods so users can check logs, or delete them and the logs are lost.

Reply:
I find that we could use `timeout` mechanism to summarize the following circumstances:

- OOM
- killed by preemption
- hang

The master will record the time when a worker get a task. Then the task will be put to doing list. The master will check doing list, if the time exceeds a value, which means the worker may fail due to some reasones, the task in doing list will be remoed to todo list.